examples/lora/lorawan/basic-demo/main.go

```diff
@@ -0,0 +1,98 @@
+// Simple code for connecting to Lorawan network and uploading sample payload
+
+package main
+
+import (
+	"time"
+
+	"tinygo.org/x/drivers/examples/lora/lorawan/common"
+	"tinygo.org/x/drivers/lora"
+	"tinygo.org/x/drivers/lora/lorawan"
+)
+
+const (
+	LORAWAN_RECONNECT_DELAY_SEC = 60
+)
+
+var (
+	radio   lora.Radio
+	session *lorawan.Session
+	otaa    *lorawan.Otaa
+
+	lorawanConnected bool
+)
+
+// loraConnect() will loop until we're connected to Lorawan network
+func loraConnect() {
```

Comment:
Just wonder why we don't just write this function in a synchronous way, e.g. 

```go
func loraConnect() error {
    start := time.Now()
    var err error
    for time.Since(start) < 30*time.Second {
        println("Starting Lorawan Join sequence")
	err = lorawan.Join(otaa, session)
        if err == nil {
            println("loraConnect: Connected !")
            return nil
        }

        println("loraConnect: Join error:", err)
	println("loraConnect: retrying")
	time.Sleep(time.Second * LORAWAN_RECONNECT_DELAY_SEC)
    }

    err = errors.New("lorawan: unable to join")
    println(err.Error())
    return err
}
```

And then call it like this:

```go
if err := loraConnect(); err != nil {
    failMessage(err)
}
```

In other words, avoid the goroutine entirely.

What do you think?

Reply:
Hi @deadprogram, 

The loraConnect() was initially design with goroutine to handle all the Lorawan connection lifecycle (machine state for join/accept/uplinks/downlinks, reconnections, retransmissions,  channel hopping, RX windows ... ). 

This way, the developper would only have to start the goroutine and interact with it through IN/OUT channel events.
(This takes advantage of goroutine/channel features, but I could be wrong ...)

That said, the current lorawan code only implements basic features ...
So we can remove the goroutine in this example.

This will make the code simple and easy to understand for newcomers and testers.

Next, I'll be happy to discuss about the correct implementation of the Lorawan machine state (goroutine or not),

I'll push the necessary changes in the next hours. 

Thanks